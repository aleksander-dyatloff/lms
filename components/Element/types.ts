import { HTMLAttributes, MouseEventHandler } from 'react'

namespace ElementComponent {
  // eslint-disable-next-line max-len
  export interface Props extends HTMLAttributes<HTMLElement> {
    readonly onClickOutside?: MouseEventHandler<HTMLElement>
  }
}

export default ElementComponent
