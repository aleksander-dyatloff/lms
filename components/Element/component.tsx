import useRefs from '@hooks/useRefs'
import { forwardRef, useEffect, useRef } from 'react'

import Wrapper from './styles'
import ElementComponent from './types'

const Element = forwardRef<HTMLElement | null, ElementComponent.Props>(({
  onClickOutside,
  onFocus,
  onClick,
  onClickCapture,
  tabIndex = (onFocus ?? onClick ?? onClickCapture) ? 0 : undefined,
  ...restProps
}, forwardedRef) => {
  const innerRef = useRef<HTMLElement>(null)

  const ref = useRefs(forwardedRef, innerRef)

  useEffect(() => {
    const handleOutsideClick: EventListener = (event) => {
      if (ref.current && !innerRef?.current?.contains(event.target as Node)) {
        onClickOutside?.(event)
      }
    }

    if (onClickOutside) {
      document.addEventListener('click', handleOutsideClick)
    }

    return () => document.removeEventListener('click', handleOutsideClick)
  }, [])

  return (
    <Wrapper
      tabIndex={tabIndex}
      {...restProps}
    />
  )
})

export default Element
