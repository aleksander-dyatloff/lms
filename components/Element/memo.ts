import { memo } from 'react'

import Element from './component'

export default memo(Element)
