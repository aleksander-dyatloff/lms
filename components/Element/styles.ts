import styled from 'styled-components'

const Wrapper = styled.div`
  // Reset styles
  padding: 0;
  margin: 0;
  font: inherit;
  letter-spacing: inherit;
  list-style: none;
  color: inherit;
  background-color: inherit;
  border: none;
  border-radius: 0;
  width: 100%;
  text-align: inherit;
  text-decoration: inherit;
  cursor: inherit;
  outline: none;
  caret-color: inherit;
`

export default Wrapper
