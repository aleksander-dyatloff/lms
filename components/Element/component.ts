import useRefs from '@hooks/useRefs'
import { ForwardedRef, forwardRef, ReactElement, useEffect, useRef } from 'react'

import { Wrapper } from './styles'
import { ElementComponent } from './types'

const Element = forwardRef<HTMLElement, ElementComponent.Props>(({
  as = 'div',
  onClickOutside,
  onFocus,
  onClick,
  onClickCapture,
  tabIndex = (onFocus ?? onClick ?? onClickCapture) ? 0 : undefined,
  ...restProps
}, forwardedRef) => {
  const innerRef = useRef<HTMLElement>(null)

  const ref = useRefs(forwardedRef, innerRef)

  useEffect(() => {
    const handleOutsideClick: EventListener = (event) => {
      if (ref.current && !innerRef?.current?.contains(event.target as Node)) {
        onClickOutside?.(event)
      }
    }

    if (onClickOutside) {
      document.addEventListener('click', handleOutsideClick)
    }

    return () => document.removeEventListener('click', handleOutsideClick)
  }, [])

  return (
    <Wrapper
      tabIndex={tabIndex}
...restProps}
    />
  )
// eslint-disable-next-line max-len
})

export default Element
