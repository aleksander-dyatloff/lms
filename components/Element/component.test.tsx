import 'jest-styled-components'

import { render } from '@testing-library/react'
import { Profiler } from 'react'

import Element from './component'

describe('Element component', () => {
  test('should match to snapshot', () => {
    const { container } = render(
      <Element>
        Click
      </Element>,
    )

    expect(container).toMatchSnapshot()
  })

  test('first render duration should match to specification', () => {
    render(
      <Profiler
        id='Element'
        onRender={(_, __, renderDuration) => {
          expect(renderDuration).toBeLessThanOrEqual(1.5)
        }}
      >
        <Element>
          Click
        </Element>
      </Profiler>,
    )
  })
})
