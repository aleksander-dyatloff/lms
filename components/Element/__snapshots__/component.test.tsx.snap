// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Element component should match to snapshot 1`] = `
.c0 {
  padding: 0;
  margin: 0;
  font: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  list-style: none;
  color: inherit;
  background-color: inherit;
  border: none;
  border-radius: 0;
  width: 100%;
  text-align: inherit;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  cursor: inherit;
  outline: none;
  caret-color: inherit;
}

<div>
  <div
    class="c0"
  >
    Click
  </div>
</div>
`;
